src/prepared_statement.c

```diff
@@ -0,0 +1,545 @@
+/*
+ * Contains code for handling prepared statement related packets.
+ *
+ * The main entrypoints are the handle_xxx_command functions. They do the
+ * translation from prepared statement names of the client to the equivalent
+ * prepared statement name on the server.
+ */
+
+#include "bouncer.h"
+
+#include <usual/crypto/csrandom.h>
+#include <usual/hashtab-impl.h>
+#include <usual/slab.h>
+
+static uint64_t next_unique_query_id;
+
+/*
+ * Like uthash HASH_DELETE, but doesn't remove item from hash table just
+ * unlink element from the doubly-linked-list.
+ */
+#define HASH_UNLINK(hh, head, delptr)                                                                   \
+	do {                                                                                                                            \
+		struct UT_hash_handle *_hd_hh_del = &(delptr)->hh;                              \
+		if (_hd_hh_del->prev != NULL) {                                                                 \
+			HH_FROM_ELMT((head)->hh.tbl, _hd_hh_del->prev)->next = _hd_hh_del->next; \
+		} else {                                                                                                                \
+			DECLTYPE_ASSIGN(head, _hd_hh_del->next);                                        \
+		}                                                                                                                               \
+		if (_hd_hh_del->next != NULL) {                                                                 \
+			HH_FROM_ELMT((head)->hh.tbl, _hd_hh_del->next)->prev = _hd_hh_del->prev; \
+		} else {                                                                                                                \
+			_hd_hh_del->tbl->tail = HH_FROM_ELMT(_hd_hh_del->tbl, _hd_hh_del->prev); \
+		}                                                                                                                               \
+	} while (0)
+
+#define HASH_FIND_UINT64(head, findint, out) \
+	HASH_FIND(hh, head, findint, sizeof(uint64_t), out)
+#define HASH_ADD_UINT64(head, intfield, add) \
+	HASH_ADD(hh, head, intfield, sizeof(uint64_t), add)
+
+/*
+ * Converts a PgParsePacket to a malloc-ed PgPreparedStatement. The
+ * PgPreparedStatement can be stored in the global prepared statement cache.
+ */
+static PgPreparedStatement *create_prepared_statement(PgParsePacket *pkt)
+{
+	PgPreparedStatement *ps = malloc(
+		sizeof(PgPreparedStatement)
+		+ pkt->query_and_parameters_len);
+	if (ps != NULL) {
```

Comment:
Why not bail here if the malloc returns null?

Reply:
Fixed.
Frankly speaking I also think that malloc failure should cause immediate bail.
But looks like pgbouncer approach is different and it tries to continue work even if there is no free memory. So OOM error is just propagated to caller.

Reply:
Yeah, agreed, not sure what the client is supposed to do about it. Just have pgbouncer bail. Things are bad if we can't even malloc those tiny structs. I'm also not a fan of all those gotos and duplicated code.

Reply:
I guess this comes from Postgres its way of handling OOMs too. I can see benefits of both ways for PgBouncer. It seems quite nice if some long running query on one connection doesn't get killed due to an OOM on another connection. But on the other hand, if you start getting OOMs on one connection you'll probably get them on many others too. 

In either case it's definitely not something we would want to change in this PR, so for now an OOM should simply result in the respective a client connection being killed.

Reply:
I did rewrite a few cases to use the early-return pattern, to make it clearer that we're bailing out of the rest of the function in case of OOM.